src/AmalgamatorSession.ts

```diff
@@ -451,6 +465,31 @@ export class AmalgamatorSession extends LoggingDebugSession {
         }
     }
 
+    protected async disassembleRequest(
+        response: DebugProtocol.DisassembleResponse,
+        args: DebugProtocol.DisassembleArguments
+    ): Promise<void> {
+        response.body = {
+            instructions: [],
+        };
+        if (args.memoryReference.indexOf('0x') === -1) {
+            const [, memoryReference] = this.disassembleHandles.get(
+                parseInt(args.memoryReference)
+            );
+            args.memoryReference = memoryReference.split(':')[1];
+            const disassemble = await this.childDaps[
+                parseInt(memoryReference.split(':')[0])
+            ].disassembleRequest(args);
+            response.body = disassemble.body;
```

Comment:
These lines are pretty much all this function needs to do, add error checking and that should be good.

Reply:
When we scroll disassembly view to a new address that don't have disassembled data, args.memoryReference will just contain address like 0x.... did not contain ":". So args.memoryReference will have two cases: 
- When we step over program:   contain `childIndex + ':' address`
- When we scroll disassembly view:  only contain `address`
Do you think about this?

Reply:
Can you provide a trace of that situation please?

Reply:
Below is the image that when I scroll the view to end. Could you please check it?
<img width="937" alt="ScrollViewToEnd" src="https://user-images.githubusercontent.com/111261098/197584665-ffe28ef6-2e2e-4427-9dac-94771a6d7563.png">

Reply:
Thanks for the screenshot. We may indeed need the workaround you are proposing in your code, but the trace would really help me understand what you are seeing.

Reply:
Below are the log file. I hope that they are helpful.
[llvm.txt](https://github.com/eclipse-cdt-cloud/cdt-amalgamator/files/9853846/llvm.txt) [RL.txt](https://github.com/eclipse-cdt-cloud/cdt-amalgamator/files/9853847/RL.txt)

Reply:
I have attached the log as your request. Please help me check it.
[amalgamator.txt](https://github.com/eclipse-cdt-cloud/cdt-amalgamator/files/9853955/amalgamator.txt)

Reply:
Here is the critical part of the log with long lines trimmed
```
[17:41:03.410 UTC] To client: {"seq":0,"type":"response","request_seq":9,"command":"stackTrace","success":true,"body":{"stackFrames":[{"id":1000,"source":{"name":"LLVM.c","path":"D:\\workspace\\Project_ARM_G4MH\\SIM\\LLVM\\src\\LLVM.c","sourceReference":0},"line":41,"column":0,"name":"main","instructionPointerReference":"1:0x000001cc"}],"totalFrames":1}}
[17:41:07.644 UTC] From client: disassemble({"memoryReference":"1:0x000001cc","offset":0,"instructionOffset":-200,"instructionCount":400,"resolveSymbols":true})
[17:41:07.781 UTC] To client: {"seq":0,"type":"response","request_seq":10,"command":"disassemble","success":true,"body":{"instructions":[{"address":"0x000001cc","instructionBytes":"b8 02","instruction":"movw\t[sp+2], ax","symbol":"main+8","location":{"name":"../src\\LLVM.c","path":"D:\\workspace\\Project_ARM_G4MH\\SIM\\LLVM\\src\\LLVM.c"},"line":41},{"address":"0x000001ce","instructionBytes":"30 0a 00","instruction":"movw\tax, #10","symbol":"main+10","location":...
[17:41:10.161 UTC] From client: disassemble({"memoryReference":"0x000003e5","offset":0,"instructionOffset":1,"instructionCount":50,"resolveSymbols":true})
```

I think this shows a limitation in the spec or the implementation in VSCode.

The adapter says that the reference is (correctly) `1:0x000001cc`, and when VSCode makes its first request it uses the `instructionOffset` to do the math by asking for `"memoryReference":"1:0x000001cc","offset":0,"instructionOffset":-200`, but on subsequent calls (when you scroll) it isn't simply adding/removing `instructionOffset` to get to the correct place.

So for the amalgamator, I agree that you will need a workaround. My guess is best workaround is that you remember the last dap index you returned and if there is no index in what you got back assume it is for the same child adapter. This will require a release note as there will be situations that the wrong thing is disassembled.

To fix the underlying issue, an issue/PR against VSCode will probably be needed.

Let me know when you are ready for me to re-review the code.

Reply:
There is other problem is that if we open disassembly view, then disconnect debug and debug again. disassembleRequest will not be called whenever open disassembly view. It means that disassembleRequest did not be call when open disassembly view after terminating debug. you can check on the log file below.
[amalgamator.txt](https://github.com/eclipse-cdt-cloud/cdt-amalgamator/files/9854036/amalgamator.txt)

Reply:
PS before you file an issue with VSCode it may be worth looking up if there are already issues about multiple memory spaces in VSCode's issue tracker.

Reply:
> There is other problem is that if we open disassembly view, then disconnect debug and debug again. disassembleRequest will not be called whenever open disassembly view. It means that disassembleRequest did not be call when open disassembly view after terminating debug. you can check on the log file below. [amalgamator.txt](https://github.com/eclipse-cdt-cloud/cdt-amalgamator/files/9854036/amalgamator.txt)

If disassembleRequest is not called at all, this is also a VSCode bug that needs to be reported (and recorded in release notes too)

Reply:
- [ ] File a bug with VSCode about multiple memory regions (i.e. the assumption that addresses are unique in a debug session)
- [ ] File a bug with VSCode about disassemble request not called at startup
- [ ] Ask me to review the code when you have it working. What you have most recently suggested looks very close so you can ask me to review that and then it will be easier to identify exactly what is left to be done (if anything)

Reply:
[Here is the code](https://github.com/microsoft/vscode/blob/ed31e676b61faf8dc188ab58873bc98c0b5a0993/src/vs/workbench/contrib/debug/browser/disassemblyView.ts#L320-L336) in VSCode that does math on the address instead of offsets to the original reference.

Reply:
Thank so much for your information